odl/tomo/backends/astra_setup.py

```diff
@@ -0,0 +1,581 @@
+    type_map_cuda = {'parallel': 'cuda',  # I
+                     'parallel3d': 'cuda3d'}  # I
+    type_map_cuda['fanflat'] = type_map_cuda['parallel']
+    type_map_cuda['fanflat_vec'] = type_map_cuda['fanflat']
+    type_map_cuda['cone'] = type_map_cuda['parallel3d']
+    type_map_cuda['parallel3d_vec'] = type_map_cuda['parallel3d']
+    type_map_cuda['cone_vec'] = type_map_cuda['cone']
+
+    # create config dict
+    proj_cfg = {}
+    if impl == 'cpu':
+        proj_cfg['type'] = type_map_cpu[proj_type][vol_interp]
+    else:  # impl == 'cuda'
+        proj_cfg['type'] = type_map_cuda[proj_type]
+    proj_cfg['VolumeGeometry'] = astra_vol_geom
+    proj_cfg['ProjectionGeometry'] = astra_proj_geom
+
+    if ndim == 2:
+        return astra.projector.create(proj_cfg)
+    else:
+        return astra.projector3d.create(proj_cfg)
+
+
+def astra_algorithm(direction, ndim, vol_id, sino_id, proj_id, impl):
+    """Create an ASTRA algorithm object to run the projector.
+
+    Parameters
+    ----------
+    direction : {'forward', 'backward'}
+        Apply the forward projection if 'forward', otherwise the backprojection
+    ndim : {2, 3}
+        Number of dimensions of the projector
+    vol_id : `int`
+        ASTRA ID of the volume data object
+    sino_id : `int`
+        ASTRA ID of the projection data object
+    proj_id : `int`
+        ASTRA ID of the projector
+    impl : {'cpu', 'cuda'}
+        Implementation of the projector
+
+    Returns
+    -------
+    id : `int`
+        ASTRA internal ID for the new algorithm structure
+    """
+    if direction not in ('forward', 'backward'):
+        raise ValueError('direction {!r} not understood.'.format(direction))
+    if ndim not in (2, 3):
+        raise ValueError('{}-dimensional projectors not supported.'
+                         ''.format(ndim))
+    if impl not in ('cpu', 'cuda'):
+        raise ValueError('implementation type {!r} not understood.'
+                         ''.format(impl))
+    if ndim is 3 and impl is 'cpu':
+        raise NotImplementedError(
+            '3d algorithms for cpu is not supported by ASTRA')
+    if proj_id is None and impl is 'cpu':
+        raise ValueError('cpu implementation requires projector ID')
+
+    algo_map = {'forward': {2: {'cpu': 'FP', 'cuda': 'FP_CUDA'},
+                            3: {'cpu': None, 'cuda': 'FP3D_CUDA'}},
+                'backward': {2: {'cpu': 'BP', 'cuda': 'BP_CUDA'},
+                             3: {'cpu': None, 'cuda': 'BP3D_CUDA'}}}
+
+    algo_cfg = {'type': algo_map[direction][ndim][impl],
+                'ProjectorId': proj_id,
+                'ProjectionDataId': sino_id}
+    if direction is 'forward':
+        algo_cfg['VolumeDataId'] = vol_id
+    else:
+        algo_cfg['ReconstructionDataId'] = vol_id
+
+    # Create ASTRA algorithm object
+    return astra.algorithm.create(algo_cfg)
+
+
+def astra_cleanup():
+    """Delete all ASTRA objects."""
```

Comment:
We should implement some minimal bookkeeping and then clear the objects specifically.